Add dialog light dismiss behavior

mfreed7 opened the pull request:
This adds a few things:
 - A `closedby` attribute that can be used to control the light dismiss behavior of dialogs:
   1. `<dialog closedby=none>` - no user-triggered closing of dialogs at all.
   2. `<dialog closedby=closerequest>` - user pressing ESC (or other close trigger) closes the dialog
   3. `<dialog closedby=any>` - user clicking outside the dialog, or pressing ESC, closes the dialog. Akin to `popover=auto` behavior.
   4. `<dialog>` - i.e. no `closedby` attribute - default/old behavior. Behaves like `closedby=closerequest` for modal dialogs, and `closedby=none` for modeless dialogs.

- This PR also (per the [request](https://github.com/openui/open-ui/discussions/950#discussioncomment-7592744)) adds a `dialog.requestClose()` which fires the `cancel` event, and then (if that isn't cancelled) fires the `close` event and closes the dialog.
- I also consolidated the "light dismiss" activities into one algorithm, which will need to be called by https://github.com/w3c/pointerevents/pull/460 if that ever gets some attention.

Some interesting related links:
- https://github.com/whatwg/html/pull/10157  - much of this PR was lifted from @lukewarlow's PR
- https://github.com/whatwg/html/issues/9373 - the WHATWG discussion
- https://github.com/openui/open-ui/issues/834 - the OpenUI discussion
- https://github.com/openui/open-ui/discussions/960 - naming discussion to get to `closedby`
- https://github.com/openui/open-ui/discussions/961 - naming discussion to get to `closerequest`

Closes https://github.com/whatwg/html/issues/9373

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
When editing this comment after the PR is created, check that PR-Preview doesn't overwrite your changes.
If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [X] At least two implementers are interested (and none opposed):
   * Chromium
   * WebKit: Presumably supportive, based on @annevk reviews. [No official position](https://github.com/WebKit/standards-positions/issues/329)
   * Gecko: @smaug---- has said he would review, so TBD. [No official position](https://github.com/mozilla/standards-positions/issues/998)
- [X] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/html/semantics/interactive-elements/the-dialog-element?label=master&label=experimental&aligned&q=dialog-closedby.tentative.html%20or%20dialog-popover-closedby-complex.tentative.html%20or%20dialog-popover-closedby-simple.tentative.html%20or%20dialog-requestclose.tentative.html
- [X] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://crbug.com/376516550
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1936940
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=284592
- [X] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/issues/37196
- [X] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10737/dnd.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/dnd.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/dnd.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )
<a href="https://whatpr.org/html/10737/dom.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/dom.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/dom.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )
<a href="https://whatpr.org/html/10737/index.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/index.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/index.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )
<a href="https://whatpr.org/html/10737/indices.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/indices.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/indices.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )
<a href="https://whatpr.org/html/10737/infrastructure.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/infrastructure.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )
<a href="https://whatpr.org/html/10737/interaction.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/interaction.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/interaction.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )
<a href="https://whatpr.org/html/10737/interactive-elements.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/interactive-elements.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/interactive-elements.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )
<a href="https://whatpr.org/html/10737/popover.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/popover.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/popover.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )
<a href="https://whatpr.org/html/10737/semantics-other.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">/semantics-other.html</a>  ( <a href="https://whatpr.org/html/10737/b4e1fb1...57df925/semantics-other.html" title="Last updated on Jan 9, 2025, 7:49 PM UTC (57df925)">diff</a> )

domenic commented:
This is exciting! Unfortunately I am going OOO for a week starting ~now, so won't get a chance to review too soon. Hopefully others will be able to help with some reviews, but I would like to get a chance to carefully review the close watcher integration, etc. before this lands.

Would you mind commenting on what approach you took to the mutability of the closedby attribute? We discussed that a bit in https://github.com/whatwg/html/issues/9373, and I think we came up with a tentative proposal, but I'm not sure it was battle-tested.

keithamus commented:
> Would you mind commenting on what approach you took to the mutability of the closedby attribute? We discussed that a bit in #9373, and I think we came up with a tentative proposal, but I'm not sure it was battle-tested.

Looks like `CloseWatcher`s get a mutable `enabled` value, which is toggled whenever `closedby` is set to the applicable values - when `false` it skips the events and returns early. IMO this seems to nicely side-skirt the stacking issues around mutation of the attribute without adding lots of complexity in either the consuming elements and/or having some messy algorithms to insert mid-way into a stack.

keithamus commented on source:
```diff
@@ -62046,7 +62164,134 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Run <span>Light dismiss open dialogs</span> with <var>event</var>.</p></li>
+
+   <li><p>Run <span>Light dismiss open popovers</span> with <var>event</var>.</p></li>
+
+   <p class="XXX"><span>Run light dismiss activities</span> will be called by the <a
+   href="https://github.com/w3c/pointerevents/pull/460">Pointer Events spec</a> when the user clicks
+   or touches anywhere on the page.</p>
+  </ol>
```
Does this want to live outside of the `<ol>`?

```suggestion
  </ol>
  
  <p class="XXX"><span>Run light dismiss activities</span> will be called by the <a
  href="https://github.com/w3c/pointerevents/pull/460">Pointer Events spec</a> when the user clicks
  or touches anywhere on the page.</p>
```

mfreed7 replied:
Yeah I wasn't sure whether inside or outside was more correct. Done.

keithamus reviewed (commented):
This looks great to me!

mfreed7 commented:
> This is exciting! Unfortunately I am going OOO for a week starting ~now, so won't get a chance to review too soon. Hopefully others will be able to help with some reviews, but I would like to get a chance to carefully review the close watcher integration, etc. before this lands.

That's totally fine - no rush. I did this one in the opposite order that I usually do: spec first, prototype after. So I'm moving on to prototype it now, which might reveal other flaws. But I thought since @lukewarlow had gotten the spec into such decent shape, that I'd clean it up and open the PR in case folks saw other flaws there that I could address in the prototype.

> Would you mind commenting on what approach you took to the mutability of the closedby attribute? We discussed that a bit in #9373, and I think we came up with a tentative proposal, but I'm not sure it was battle-tested.

As @keithamus said, I just kept what @lukewarlow did in the original PR, and used a new `enabled` state on `CloseWatcher`. I haven't implemented that yet, but it seems reasonably straightforward. But let me know if you see any flaws.

annevk commented on source:
```diff
@@ -86342,10 +86601,6 @@ dictionary <dfn dictionary>DragEventInit</dfn> : <span>MouseEventInit</span> {
    </li>
   </ol>
 
-  <p class="XXX"><span>Light dismiss open popovers</span> will be called by the <a
-  href="https://github.com/w3c/pointerevents/pull/460">Pointer Events spec</a> when the user clicks
-  or touches anywhere on the page.</p>
-
```
This should be restored.

mfreed7 replied:
You'd prefer it to exist in two places? (I moved it up next to the (one) place where the pointer events spec will call into, at `<dfn>run light dismiss activities</dfn>`.)

I can add it here also, just double-checking.

annevk replied:
No that works. Thanks!

annevk commented on source:
```diff
@@ -61670,6 +61673,39 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+     <td>No user actions automatically close the dialog.
+  </table>
+
+  <p>The <code data-x="attr-dialog-closedby">closedby</code> attribute's <i data-x="invalid value
+  default">invalid value default</i> and <i data-x="missing value default">missing value
+  default</i> are both the <dfn data-x="attr-closedby-auto-state">Auto</dfn> state. The <span
+  data-x="attr-closedby-auto-state">auto</span> state matches <span
+  data-x="attr-closedby-closerequest-state">closerequest</span> when the element is modal;
+  otherwise <span data-x="attr-closedby-none-state">none</span>.</p>
```
Also, perhaps the statement of fact is better as a note?

mfreed7 replied:
> There's very inconsistent spelling of states here. They should match the capitalization and words of how they are introduced. And should typically be followed by the word state (though that's not linked).

Yep, I agree. I fixed them all up, I think. Please let me know if you see any others I missed. All are now lowercase.

> Also, perhaps the statement of fact is better as a note?

Sorry - can you help me understand this? Which "statement of fact"?

annevk commented on source:
```diff
@@ -61670,6 +61673,39 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+
+  <p>The <code data-x="attr-dialog-closedby">closedby</code> attribute's <i data-x="invalid value
+  default">invalid value default</i> and <i data-x="missing value default">missing value
+  default</i> are both the <dfn data-x="attr-closedby-auto-state">auto</dfn> state. The <span
+  data-x="attr-closedby-auto-state">auto</span> state matches <span
+  data-x="attr-closedby-closerequest-state">close request</span> state when the element is modal;
+  otherwise the <span data-x="attr-closedby-none-state">none</span> state.</p>
```
This statement about the auto state is better done as a note I think. At least presumably the actual requirements that lead to this are elsewhere.

annevk commented on source:
```diff
@@ -61747,6 +61790,27 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+
+     <li><p><i data-x="create-close-watcher-enabled">enabled</i> being true if <span>this</span>'s
+     <code data-x="attr-dialog-closedby">closedby</code> attribute state is <span
+     data-x="attr-closedby-any-state">Any</span> or <span
+     data-x="attr-closedby-closerequest-state">Close Request</span>; otherwise false.</p></li>
```
You didn't change the casing consistently it seems. I have to say I actually prefer this casing as it makes it clearer what the states are relative to the values.

annevk replied:
Close Request seems good. That matches what we use for `<input type>` and should use elsewhere imo.

annevk commented on source:
```diff
@@ -62024,8 +62111,42 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   return.</p></li>
+
+   <li><p>If <var>oldValue</var> and <var>value</var> are in the same <span
+   data-x="attr-dialog-closedby">state</span>, then return.</p></li>
```
This reads a little weird. Could you file a follow-up issue to improve this as it's not immediately clear to me what the correct wording would be and it does seem like we want to be able to do this.

mfreed7 replied:
https://github.com/whatwg/html/issues/10761

annevk commented on source:
```diff
@@ -62024,8 +62111,42 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p><span>Assert</span>: <var>element</var>'s <span data-x="dialog-close-watcher">close
+   watcher</span> is not null.</p></li>
+
+   <li><p>If <var>value</var> is in the <span data-x="attr-closedby-any-state">Any</span> state,
```
Here I think we can talk about the state of element's closedby attribute, no?

mfreed7 replied:
Just to make sure I understand this comment: these steps run *after* the attribute has been set? So it's correct to just check `closedBy`? I've done that, assuming the answer is "yes". Still feels a little weird not to use `<var>value</var>`.

mfreed7 replied:
I think this comment is likely outdated. I'll resolve it, but if I'm wrong, feel free to re-open.

annevk commented on source:
```diff
@@ -62046,7 +62167,134 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
 
+  <h4><dfn>Dialog light dismiss</dfn></h4>
+
+  <p class="note">"Light dismiss" means that clicking outside of a dialog whose <code
```
dialog element* (also needs code)

mfreed7 replied:
Done. The third one is questionable: `<code>dialog</code>s`

annevk commented on source:
```diff
@@ -62046,7 +62167,134 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+
+   <li><p>Let <var>document</var> be <var>target</var>'s <span>node document</span>.</p></li>
+
+   <li><p>If <var>document</var>'s <span>light dismissible dialog list</span> is empty, then
```
xref is empty

mfreed7 replied:
For `<span>node document</span>`? It seems correctly linked in the preview, right?

annevk replied:
No, "is empty".

mfreed7 replied:
Ahh that one. Thanks, done.

annevk commented on source:
```diff
@@ -62046,7 +62167,134 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>dialogs</var> be « ».</p></li>
+
+   <li><p><span data-x="list iterate">For each</span> <code>Element</code> <var>element</var> in
+   <var>document</var>: if <var>element</var> is a <code data-x="HTMLDialogElement">dialog element</code>,
```
Order? I'm also not sure if "element in document" works. "element in document's descendants" maybe?

mfreed7 replied:
Done.

annevk commented on source:
```diff
@@ -62046,7 +62167,134 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+     return <var>currentNode</var>.</p></li>
+
+     <li><p>Set <var>currentNode</var> to <var>currentNode</var>'s parent in the <span>flat
+     tree</span>.</p></li>
```
Should this use the flat tree or the composed tree?

mfreed7 replied:
I was thinking it was best to stay parallel to Popover and use flat tree:

https://html.spec.whatwg.org/#nearest-inclusive-open-popover

annevk commented on source:
```diff
@@ -61670,6 +61673,39 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+     <th>Brief description
+   <tbody>
+    <tr>
+     <td><dfn attr-value for="dialog/closedby"><code data-x="attr-closedby-any">any</code></dfn>
```
This should be `attr-dialog-closedby-` here and elsewhere.

annevk reviewed (commented):
I also just noticed you forgot to update the indexes. There's a checklist at the top of `source` for what to change when you add a new attribute.

mfreed7 commented:
> I also just noticed you forgot to update the indexes. There's a checklist at the top of `source` for what to change when you add a new attribute.

Done.

mfreed7 commented:
> @domenic and @annevk heads-up, please hold off on reviewing this PR for just a bit. As I'm implementing, I've realized a few things (e.g. [#10164 (comment)](https://github.com/whatwg/html/issues/10164#issuecomment-2479792309)) that will need to change things a bit. I'm going to get it working in code first, and then come back to this PR to line things up. As-is, it's a little broken for non-modal dialogs at least.

Alright, my prototype code will still need a review, but I've updated the code, added more comprehensive tests, and updated the spec PR. Seems good to me, but let me know what you think.

domenic commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  a <code>PointerEvent</code> <var>event</var>:</p>
+
+  <ol>
+   <li><p>If <var>node</var> is an <span data-x="HTMLDialogElement">HTML dialog element</span>,
```
```suggestion
   <li><p>If <var>node</var> is a <code>dialog</code> element,
```

domenic commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  <ol>
+   <li><p>If <var>node</var> is an <span data-x="HTMLDialogElement">HTML dialog element</span>,
+   <var>node</var> has an <code data-x="attr-dialog-open">open</code> attribute, <var>node</var>'s
+   <span>is modal</span> flag is true, and <var>event</var>'s <code data-x="">clientX</code> and
```
It'd be ideal to cross-link these to https://drafts.csswg.org/cssom-view/#dom-mouseevent-clientx .

domenic commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>clickedDialog</var> be the result of running <span>nearest inclusive open
+   dialog</span> given <var>node</var>.</p></li>
+
+   <li><p>Return <var>clickedDialog</var>.</p></li>
```
Combine these two steps:

> Return the nearest inclusive open dialog given _node_.

domenic commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Return <var>clickedDialog</var>.</p></li>
+  </ol>
+
+  <p>To get the <dfn id="light-dismissible-dialog-list">light dismissible dialog list</dfn> for a
```
This id="" should not be necessary.

mfreed7 replied:
Done.

domenic commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>dialogs</var> be « ».</p></li>
+
+   <li><p><span data-x="list iterate">For each</span> <code>Element</code> <var>element</var> in
+   <var>document</var>'s descendants, in <span>shadow-including tree order</span>: if <var>element</var>
```
I think this should be shadow-including descendants, and link to https://dom.spec.whatwg.org/#concept-shadow-including-descendant ?

mfreed7 replied:
Done.

domenic commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+
+   <li><p><span data-x="list iterate">For each</span> <code>Element</code> <var>element</var> in
+   <var>document</var>'s descendants, in <span>shadow-including tree order</span>: if <var>element</var>
+   is a <code data-x="HTMLDialogElement">dialog element</code> and <var>element</var> has the <code
```
```suggestion
   is a <code>dialog</code> element and <var>element</var> has the <code
```

mfreed7 replied:
Done.

domenic commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+
+    <ol>
+     <li><p>If <var>currentNode</var> is an <span data-x="HTMLDialogElement">HTML dialog
+     element</span> and <var>currentNode</var> has an <code data-x="attr-dialog-open">open</code>
```
Same issue referencing dialog element.

mfreed7 replied:
Done.

domenic commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   </li>
+
+   <li><p>Return null.</p></li>
+  </ol>
```
Optional: "nearest inclusive open dialog" is only used once, in "topmost clicked dialog", and "topmost clicked dialog" is only used once, in "light dismiss open dialogs". It might be nicer to inline them, but, I also appreciate the benefits of factoring algorithms out for clarity, so I'm OK either way. (Maybe the intermediate of inlining "nearest inclusive open dialog" would be best?)

mfreed7 replied:
Yeah, good point. I like the intermediate suggestion. It leaves `<span>topmost clicked dialog</span>` which is fairly parallel to `<span>topmost clicked popover</span>`. But I inlined "nearest inclusive open dialog" since that was just a quick parent walk anyway.

domenic commented on source:
```diff
@@ -82964,6 +83253,9 @@ body { display:none }
    return.</p></li>
 
+   <li><p>If <var>closeWatcher</var>'s <span data-x="close-watcher-enabled">enabled</span> is false,
+   then return true.</p></li>
```
```suggestion
   then return.</p></li>
```

mfreed7 replied:
Done.

domenic reviewed (commented):
Sorry for the big delay! Overall this looks good, with the only really substantial suggestion being to move to making enabled an algorithm to avoid the attribute changed steps.

Do you have a link for the tests? I've found that writing comprehensive-feeling tests for close watcher stuff is hard (as you want to test at least two, sometimes more, nested). And then you have all the permutations of attribute values, what if someone modifies the attribute while it's closed, etc. In my experience, without tests it's quite difficult to make sure the behavior is reasonable.

For example, it'd be worth testing (maybe interactively, instead of in WPTs, at least at first) that you get reasonable user behavior in situations like one closedby=none dialog sandwiched in the middle of two closedby=any dialogs.

mfreed7 commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+     </li>
+
+     <li>
+      <p>For each <var>dialog</var> in <var>dialogList</var>, run these substeps:</p>
```
Done.

mfreed7 commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+      <p>For each <var>dialog</var> in <var>dialogList</var>, run these substeps:</p>
+
+      <ol>
+       <li><p>If <var>dialog</var> is <var>ancestor</var>, continue.</p></li>
```
Done.

mfreed7 commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+       <li><p>If <var>dialog</var>'s <span>closedby state</span> is not
+       <span data-x="attr-dialog-closedby-any-state">Any</span>, continue.</p></li>
+
+       <li><p>Assert <var>dialog</var>'s <span data-x="dialog-close-watcher">close watcher</span> is not
```
Done.

mfreed7 reviewed (commented):
Thanks for the very thorough review! I think I got everything, but let me know what I missed or did incorrectly.

mfreed7 commented:
> Sorry for the big delay! Overall this looks good, with the only really substantial suggestion being to move to making enabled an algorithm to avoid the attribute changed steps.

No problem, and thanks again!

> Do you have a link for the tests? I've found that writing comprehensive-feeling tests for close watcher stuff is hard (as you want to test at least two, sometimes more, nested). And then you have all the permutations of attribute values, what if someone modifies the attribute while it's closed, etc. In my experience, without tests it's quite difficult to make sure the behavior is reasonable.
> For example, it'd be worth testing (maybe interactively, instead of in WPTs, at least at first) that you get reasonable user behavior in situations like one closedby=none dialog sandwiched in the middle of two closedby=any dialogs.

So I've [written tests](https://wpt.fyi/results/html/semantics/interactive-elements/the-dialog-element?label=master&label=experimental&aligned&q=dialog-closedby.tentative.html%20or%20dialog-popover-closedby-complex.tentative.html%20or%20dialog-popover-closedby-simple.tentative.html%20or%20dialog-requestclose.tentative.html) for several levels of "nesting", including popover/dialog/popover/dialog with mixed modal/modeless dialogs (that's the `-complex` test). I also added just popover/dialog with different values of `closedby` (that's the `-simple` test). But I didn't include `closedby` permutations in the complex test. I can do that, if you think it'd be helpful.

domenic commented:
Did you see my comments on https://github.com/whatwg/html/issues/9373#issuecomment-2490229365? I guess we should figure out the desired behavior for nested dialogs over there...

mfreed7 commented:
> Did you see my comments on [#9373 (comment)](https://github.com/whatwg/html/issues/9373#issuecomment-2490229365)? I guess we should figure out the desired behavior for nested dialogs over there...

Nope, missed it until now. Thanks. I'll come back here when we have more clarity on the behavior.

mfreed7 commented on source:
```diff
@@ -61590,9 +61592,11 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>] undefined <span data-x="dom-dialog-showModal">showModal</span>();
-  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-close">close</span>(optional DOMString returnValue);
+  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-close">close</span>(optional DOMString returnValue = null);
+  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-request-close">requestClose</span>(optional DOMString returnValue = null);
```
I see a thumbs-up from @domenic so I'll just go back to what I had before the suggestion to move to this.

mfreed7 commented on source:
```diff
@@ -61999,6 +62081,30 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Assert: <var>this</var>'s <span data-x="dialog-close-watcher">close watcher</span>
+   is not null.</p></li>
+
+   <li><p>If <var>returnValue</var> is not given, then set it to null.</p></li>
```
Undone, per the above.

mfreed7 commented on source:
```diff
@@ -62149,8 +62289,137 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p><span>Assert</span>: <var>event</var>'s <code
+   data-x="dom-Event-isTrusted">isTrusted</code> attribute is true.</p></li>
+
+   <li><p><span>Assert</span>: <var>event</var> is a <code>PointerEvent</code>.</p></li>
```
Fair enough. I made both this and `light dismiss open popovers` take `PointerEvent`s. That should work with https://github.com/w3c/pointerevents/pull/460, which still needs to be updated, but should only be passing `PointerEvent`s as-is.

mfreed7 commented:
> > Did you see my comments on [#9373 (comment)](https://github.com/whatwg/html/issues/9373#issuecomment-2490229365)? I guess we should figure out the desired behavior for nested dialogs over there...
> 
> Nope, missed it until now. Thanks. I'll come back here when we have more clarity on the behavior.

Ok, the meeting to discuss the behavior is tomorrow, but there's already been some discussion on the [OpenUI issue](https://github.com/openui/open-ui/issues/1128), plus the your [comment on the WHATWG issue](https://github.com/whatwg/html/issues/9373#issuecomment-2490229365). Given what I've heard so far, I'm fairly confident in the outcome; so I just went ahead and modified this spec PR to only close the topmost dialog when a light-dismiss click happens. I also prototyped it, and I agree that it's much better and more predictable for the user. So we'll see what happens at the meeting, but in the meantime, you can see the new proposed spec text here.

lukewarlow commented on source:
```diff
@@ -62127,8 +62251,28 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
   attribute set this element to the currently <span>focused</span> element during the <span
   data-x="show popover">show popover algorithm</span>.</p>
 
+  <p>The following <span data-x="concept-element-attributes-change-ext">attribute change
```
These changed steps aren't needed anymore as they don't do anything.

lukewarlow replied:
Though I do wonder if it's better to just give close watcher an enabled boolean (rather than a getter) and then to update dialogs enabled boolean within these attribute changed steps? That's what Chrome actually does and feels neater?

mfreed7 replied:
Good catch - this is indeed not needed. The spec and the implementation don't have to match, and I do actually agree with @domenic's suggestion here, since it makes the timing more clear, I think.

lukewarlow commented on source:
```diff
@@ -61738,6 +61776,27 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+    handlers, by having all non-<span data-x="close request">close request</span> closing
+    affordances call this method.</p>
+
+    <p>Note that this method ignores the <code data-x="attr-dialog-closedby">closedby</code>
```
Is this note wrong? requestClose throws when closedby is none?

mfreed7 replied:
Hmm - the note says "this method ignores the closedby attribute". Help me understand why you're saying it says it throws? (It shouldn't throw, and should be independent of `closedby`.

mfreed7 replied:
Pointed out below. I think this comment is still correct, and the algo below was wrong.

lukewarlow commented on source:
```diff
@@ -61999,6 +62091,30 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <li><p>If <var>this</var> does not have an <code data-x="attr-dialog-open">open</code>
+   attribute, then return.</p></li>
+
+   <li><p>If <var>this</var>'s <span>computed closed-by state</span> is
```
@mfreed7 here it checks closedby state and throws if it's None.

mfreed7 replied:
Gah. Sorry - forgot about that. And I'm really not sure why I had that in there, perhaps other than to ensure there's a close watcher attached? I've removed that, and left in the Assert. I'll do the same in code and see if I hit that Assert anywhere. But for now, done, and thanks for catching this.

mfreed7 replied:
Alright, the assert seems fine, but this did point out another thing that makes the closewatcher `enabled` state a bit messier. Still works, but I've had to add another flag to `<dialog>`. @domenic let me know if you still like the `create-close-watcher-getEnabledState` construction, or if you'd rather go back to just a bool flag.

domenic replied:
I like the current state. requestClose() makes things messy, but it's clear from the names that this is requestClose()'s "fault", so it's pretty understandable I think.

annevk commented on source:
```diff
@@ -61844,12 +61905,12 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
    <span>"<code>InvalidStateError</code>"</span> <code>DOMException</code>.</p></li>
 
+   <li><p>If <span>this</span>'s <span>node document</span> is not <span>fully active</span>, then
+   throw an <span>"<code>InvalidStateError</code>"</span> <code>DOMException</code>.</p></li>
```
Why was this moved? It's not observable so I'd rather keep it the same as it was.

annevk commented on source:
```diff
@@ -61927,6 +61972,54 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+
+  <ol>
+   <li>
+    <p>If the state of <var>dialog</var>'s <code data-x="attr-dialog-closedby">closedby</code>
+    attribute is <span data-x="attr-dialog-closedby-auto-state">Auto</span>, then:</p>
```
```suggestion
    <p>If the state of <var>dialog</var>'s <code data-x="attr-dialog-closedby">closedby</code>
    attribute is <span data-x="attr-dialog-closedby-auto-state">Auto</span>:</p>
```
(if with nested substeps doesn't use then anymore)

domenic commented on source:
```diff
@@ -61999,6 +62092,32 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
   </ol>
 
+  <p>The <dfn method for="HTMLDialogElement"><code
+  data-x="dom-dialog-request-close">requestClose(<var>returnValue</var>)</code></dfn> method steps
```
```suggestion
  data-x="dom-dialog-requestClose">requestClose(<var>returnValue</var>)</code></dfn> method steps
```

domenic commented on source:
```diff
@@ -62113,9 +62234,18 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  <p>Each <code>dialog</code> element has a <dfn>request close return value</dfn>, which is a
+  string, initially null.</p>
+
+  <p>Each <code>dialog</code> element has an <dfn>enable closewatcher for requestclose</dfn> flag,
```
Nit: include spaces consistently, i.e. "enable close watcher for request close".

Alternate naming scheme for both this and "request close return value" would be to spell out `requestClose()`, as something like:

```html
<dfn>enable close watcher for <code data-x="dom-dialog-requestClose">requestClose()</code></dfn>

<span>enable close watcher for <code data-x="dom-dialog-requestClose">requestClose()</code></span>
```

I slightly prefer the latter, if you can pull it off.

domenic replied:
flag -> boolean

domenic commented on source:
```diff
@@ -62149,8 +62284,124 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+     <li><p>If <var>topmostDialog</var>'s <span>computed closed-by state</span> is not <span
+     data-x="attr-dialog-closedby-any-state">Any</span>, then return.</p></li>
+
+     <li><p><span>Assert</span> <var>topmostDialog</var>'s <span
```
```suggestion
     <li><p><span>Assert</span>: <var>topmostDialog</var>'s <span
```

domenic commented on source:
```diff
@@ -62149,8 +62284,124 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+  or touches anywhere on the page.</p>
+
+  <p>To find the <dfn>topmost clicked dialog</dfn>, given a <code>Node</code> <var>node</var> and
+  a <code>PointerEvent</code> <var>event</var>:</p>
```
I think this algorithm can just take the event and grab the node from the target. That seems a bit clearer to me.

domenic commented on source:
```diff
@@ -62149,8 +62284,124 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   <var>element</var> to <var>dialogs</var>.</p></li>
+
+   <li><p>Return <var>dialogs</var>.</p></li>
+  </ol>
```
I'm confused at the relationship between:

- light dismissible dialog list
- topmost clicked dialog
- the check in step 1 of "topmost clicked dialog"
- the check in step 3.1 of "topmost clicked dialog"

I haven't tried to fully understand the pointerdown-pointerup-ancestor-tracking behavior of light dismiss (which I see popover also has). But I worry about the inconsistencies here. Like, why are we only checking "is modal" and the clientX/clientY in step 1 of "topmost clicked dialog", and not anywhere else? Why does popover not have an analog of "light dismissible dialog list"? What is the relationship between "topmost clicked dialog" and the last element of "light dimissible dialog list" (consulted in light dismiss step 7.5)? Why is the order of "light dismissible dialog list" shadow-including tree-order but the order of "topmost clicked dialog" is flat tree order?

I apologize if these things are clear to someone who fully understands the pointerdown-pointerup-ancestor tracking. But maybe any explanation you can give me here will be something we can turn into a `<div class="note">` to help other spec readers.

mfreed7 replied:
No problem - thanks for asking. Let me see if I can help, and then we can figure out what a Note should say.

> * light dismissible dialog list

This just makes it possible to grab the topmost (last-opened) dialog, which is the only one that might be light dismissed. It keeps track of all open dialogs. It used to include more about the `closedby` attribute, but it no longer does, so we could rename it to "open dialog list" or similar?

> * topmost clicked dialog

This does two things:
 1. Find the nearest open dialog that the user clicked on, if any.
 2. Distinguishes between clicks on a `<dialog>` and clicks on the `::backdrop` of a `<dialog>`. (This one has a `Note` but it could be clarified maybe?)

> * the check in step 1 of "topmost clicked dialog"

This is step 2 just above ("distinguishes..."). Clicking on the `::backdrop` of a modal `<dialog>` results in a `PointerEvent` with a `target` equal to the `<dialog>` itself. The `clientX/Y` checks figure out whether the click was on the `<dialog>` or on the `::backdrop`. It returns `null` if it was on the `::backdrop` since that's a "click outside" the dialog.

> * the check in step 3.1 of "topmost clicked dialog"

This just returns the nearest ancestor open dialog.

> I haven't tried to fully understand the pointerdown-pointerup-ancestor-tracking behavior of light dismiss (which I see popover also has). But I worry about the inconsistencies here. 

Right - this behavior is modeled after popover's similar logic. The reason for the pointerdown/pointerup checks are:
 1. The a11y folks say that no destructive or hard-to-reverse actions should happen on pointerdown, because a user might accidentally click and needs a way to "escape" - by moving the mouse outside the thing they clicked on and *then* letting go. The same is true of e.g. `<button>`'s.
 2. (Different from buttons) A user might also want to text-select the contents of something outside the popover/dialog, by doing pointerdown, drag, pointerup. If that was entirely outside the popover/dialog, then it should close the popover/dialog. If, however, the user is dragging to or from the popover/dialog (like starting in the page and ending within the popover/dialog), then that shouldn't light dismiss, so that their text selection remains and the popover/dialog doesn't abruptly close with part of their selection inside it. You can try this out today with Popover - the intention is to roughly follow that model for `<dialog>`.

> Like, why are we only checking "is modal" and the clientX/clientY in step 1 of "topmost clicked dialog", and not anywhere else?

That step is just about `::backdrop`, which only exists for modal dialogs. Hence the "is modal" check.

> Why does popover not have an analog of "light dismissible dialog list"?

Popover has the equivalent concept, which is `<span>showing auto popover list</span>` (and [soon](https://github.com/whatwg/html/pull/9778), `<span>showing hint popover list</span>`).

> What is the relationship between "topmost clicked dialog" and the last element of "light dismissible dialog list" (consulted in light dismiss step 7.5)?

So "topmost clicked dialog" is the nearest open dialog that the user clicked on, or `null` if they didn't click on a dialog. Note that this might not be the "topmost" dialog - I see the confusion here. Perhaps "topmost clicked dialog" should be something like "nearest clicked dialog" - would that make it clearer? (Note: if so, we should probably also rename "topmost clicked popover" to "nearest clicked popover", but I don't know how we go about such renaming in specs. Perhaps it breaks links?) The last element of "light dismissible dialog list" is indeed the "topmost" open dialog.

> Why is the order of "light dismissible dialog list" shadow-including tree-order but the order of "topmost clicked dialog" is flat tree order?

Oh shoot! I see one bit of confusion. My PR is wrong here - in the Chromium [implementation](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_dialog_element.cc;l=316;drc=d8a8f83689fe2481ccbdbcb998336273b8cea4cf) (which I believe is the desired/correct behavior) the "light dismissible dialog list" list is stateful - it's the list of open dialogs, in the **order they were opened**. This should not be in tree order at all here - I'll try to fix this. I think it'll require keeping track of that state explicitly. For popover, we don't keep that state, [we re-use the top layer's existing list of elements, and filter it](https://html.spec.whatwg.org/#auto-popover-list). I was trying (but failed) to keep this parallel to that. Because modeless dialogs aren't in the top layer, I can't re-use it at all. 

> I apologize if these things are clear to someone who fully understands the pointerdown-pointerup-ancestor tracking. But maybe any explanation you can give me here will be something we can turn into a `<div class="note">` to help other spec readers.

Nope, all good questions, including some "bugs".

mfreed7 replied:
Ok, followup to this. I made these changes. I'll leave anything else until you reply here.

1. Added `open dialogs list` to `document`, and removed the previous `light dismissible dialog list` concept entirely. This is both a rename and a re-definition, and it now mirrors Chromium's implementation.
2. Rename `topmost clicked popover` to `nearest clicked dialog`. I did not rename the equivalent popover concept.

domenic replied:
Something about this clicked for me the second time I read through it. I can't point to anything specific in your explanation that helped enough that it needs to be in a note, so maybe the bugfix and renames were enough here. I'm happy to resolve this.

domenic reviewed (commented):
Overall looking good, except I don't understand the light dismiss targeting behavior and I think I need you to explain it to me.

domenic commented:
Oh, I spotted another issue, around the use of "request to close". I will discuss on https://github.com/openui/open-ui/issues/1128.

mfreed7 commented:
> Oh, I spotted another issue, around the use of "request to close". I will discuss on [openui/open-ui#1128](https://github.com/openui/open-ui/issues/1128).

Note that my most recent commit does *not* yet handle this. I'll work on that now.

mfreed7 commented:
> > Oh, I spotted another issue, around the use of "request to close". I will discuss on [openui/open-ui#1128](https://github.com/openui/open-ui/issues/1128).
> 
> Note that my most recent commit does _not_ yet handle this. I'll work on that now.

Alright, [latest commit](https://github.com/whatwg/html/pull/10737/commits/6aa73868d7cbf24a35f580ac703a77bdd8315bde) adds something - easier than I thought. I'll try the equivalent thing in Chromium and make sure it looks functional. But seems like it?

domenic commented on source:
```diff
@@ -61878,6 +61881,13 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
 
+   <li><p><span>Assert</span>: <span>this</span>'s <span>node document</span>'s <span>open
+   dialogs list</span> does not <span data-x="list contains">contain</span>
+   <span>this</span>.</p></li>
```
I think this assert fails if you do

```js
dialogEl.show();
dialogEl.removeAttribute("open");
dialogEl.show();
````

https://github.com/whatwg/html/pull/10124 will fix this, but I'm hesitant to block this PR on that.

domenic replied:
Yeah, as long as you can verify the assert is not load-bearing, then just removing it for now seems reasonable.

mfreed7 replied:
Well, it's load bearing in the case you bring up above - removing the `open` attribute. But the spec isn't super clear about that now anyway, so I guess it only bears code-implementation weight, not spec weight. LMK if that makes sense.

domenic commented on source:
```diff
@@ -83166,7 +83448,7 @@ dictionary <dfn dictionary>CloseWatcherOptions</dfn> {
   data-x="dom-CloseWatcher-requestClose">requestClose()</code></dfn> method steps are to <span
   data-x="close-watcher-request-close">request to close</span> <span>this</span>'s <span>internal
-  close watcher</span>.</p>
+  close watcher</span> with false.</p>
```
I think this should also be true, to match `<dialog>`. (This may require updating some tests to use the `sendCloseRequest()` helper instead of using `requestClose()`.)

mfreed7 replied:
See my comment here https://github.com/openui/open-ui/issues/1128#issuecomment-2537377041. I'm a little worried about compat, and I'd prefer to split out this change from the dialog change. If that works for you?

mfreed7 replied:
... per my next comment https://github.com/openui/open-ui/issues/1128#issuecomment-2539519339 I've made this change. I'll leave this comment open to ensure you notice it.

mfreed7 commented on source:
```diff
@@ -62149,7 +62291,144 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
+   </li>
+  </ol>
+
+  <p>To <dfn>run light dismiss activities</dfn>, given an <code>Event</code> <var>event</var>:</p>
```
I'm going to close this comment. LMK if you want this chunk moved.

domenic reviewed (approved):
Provisionally approving under the assumption the broken assert is removed. Great stuff!

This will still need the rest of the checkboxes filled out, and I'm going OOO for the rest of the year. But I might check in since there's probably not much left to do here.

One thing I haven't checked recently is the WPTs. I'd encourage adding lots and lots, e.g. for the stuff we changed recently around allowing cancelable `requestClose()`, tree order vs. flat tree order vs. dialog open order, etc. This stuff is just so subtle you can never have enough tests.

annevk commented on source:
```diff
@@ -61590,9 +61595,11 @@ interface <dfn interface>HTMLDialogElement</dfn> : <span>HTMLElement</span> {
   [<span>CEReactions</span>] undefined <span data-x="dom-dialog-show">show</span>();
   [<span>CEReactions</span>] undefined <span data-x="dom-dialog-showModal">showModal</span>();
   [<span>CEReactions</span>] undefined <span data-x="dom-dialog-close">close</span>(optional DOMString returnValue);
+  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-requestclose">requestClose</span>(optional DOMString returnValue);
```
Noticed this minor thing while skimming the latest commit (needs changing throughout):
```suggestion
  [<span>CEReactions</span>] undefined <span data-x="dom-dialog-requestClose">requestClose</span>(optional DOMString returnValue);
```

mfreed7 replied:
Ick, that was in three places, and the correct capitalization in another four. Fixed.

smaug---- commented:
FWIW, I started to review the PR but need to still think whether there are any unexpected behaviors related to click handling and targeting.

mfreed7 commented:
> Provisionally approving under the assumption the broken assert is removed. Great stuff!

Thanks for the detailed review!

> This will still need the rest of the checkboxes filled out, and I'm going OOO for the rest of the year. But I might check in since there's probably not much left to do here.

Ok, sounds good. No rush - January is ok by me.


> One thing I haven't checked recently is the WPTs. I'd encourage adding lots and lots, e.g. for the stuff we changed recently around allowing cancelable `requestClose()`, tree order vs. flat tree order vs. dialog open order, etc. This stuff is just so subtle you can never have enough tests.

Yeah, I'll give it more thought. I did add some complex cases, but I'll see what else I can add. I agree it's complicated and needs good testing.

mfreed7 commented:
> FWIW, I started to review the PR but need to still think whether there are any unexpected behaviors related to click handling and targeting.

Great, thanks for reviewing. Let me know what you find.

lukewarlow commented:
Fwiw regarding implementor interest I raised https://github.com/mozilla/standards-positions/issues/998 and https://github.com/WebKit/standards-positions/issues/329

mfreed7 commented:
> Fwiw regarding implementor interest I raised [mozilla/standards-positions#998](https://github.com/mozilla/standards-positions/issues/998) and [WebKit/standards-positions#329](https://github.com/WebKit/standards-positions/issues/329)

Oh, thanks! I didn't realize those had been filed. I added them to the OP block.

domenic commented:
Back from vacation, and merged! Please remember to send a PR renaming the tests from .tentative.

stof commented:
Would `closedBy=any` also close when clicking on the backdrop pseudo-element or is it not considered a click outside the dialog ?

lukewarlow commented:
> Would closedBy=any also close when clicking on the backdrop pseudo-element or is it not considered a click outside the dialog ?

Yes, that closes the dialog. Here's a demo you can play around with in Chrome canary (w/ experimental web flag): https://jsfiddle.net/vco1mdth/

mfreed7 commented:
> Back from vacation, and merged! Please remember to send a PR renaming the tests from .tentative.

Welcome back, and thanks!
